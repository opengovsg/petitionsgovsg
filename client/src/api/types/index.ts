export * from './post'
export * from './signature'
export * from './addressee'
export * from './auth'
