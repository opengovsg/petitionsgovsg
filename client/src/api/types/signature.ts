import { BaseModelParams, MessageResponse } from './common'

export type BaseSignatureDto = BaseModelParams & {
  comment: string | null
  userId: number
  postId: number
  fullname: string | null
}

export type CreateSignatureReqDto = { text: string }

// Returns signature ID
export type CreateSignatureResDto = MessageResponse & { data: number }
