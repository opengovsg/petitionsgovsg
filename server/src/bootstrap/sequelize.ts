import { Sequelize } from 'sequelize'

import {
  defineSignature,
  definePost,
  defineAddressee,
  defineSubscription,
} from '@/models'
import { dbConfig } from './config/database'

export const sequelize = new Sequelize({ ...dbConfig, logging: false })

export const { Addressee } = defineAddressee(sequelize)
export const { Post } = definePost(sequelize, { Addressee })
export const Signature = defineSignature(sequelize, { Post })
export const Subscription = defineSubscription(sequelize, { Post })

export default sequelize
