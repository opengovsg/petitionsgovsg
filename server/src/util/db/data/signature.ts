export const mockSignature = {
  id: 4,
  hashedUserSgid:
    '$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW',
  fullname: null,
  comment: 'I want to support this cause.',
  postId: 'af7b6aa9-7c6e-4ea5-b536-4e55bc506847',
  createdAt: '2022-01-01T00:00:00+00:00',
  updatedAt: '2022-01-01T00:00:00+00:00',
  type: 'support',
}

export const mockSignatures = [
  {
    id: 1,
    hashedUserSgid: '1234',
    fullname: null,
    comment: null,
    postId: 'af7b6aa9-7c6e-4ea5-b536-4e55bc506847',
    type: 'support',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
  {
    id: 2,
    hashedUserSgid: '5678',
    fullname: 'John Doe',
    comment: null,
    postId: 'af7b6aa9-7c6e-4ea5-b536-4e55bc506847',
    type: 'support',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
  {
    id: 3,
    hashedUserSgid: '8765',
    fullname: null,
    comment: 'I support this cause!',
    postId: 'af7b6aa9-7c6e-4ea5-b536-4e55bc506847',
    type: 'support',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
]
