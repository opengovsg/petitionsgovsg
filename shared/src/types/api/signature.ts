export type CreateSignatureReqDto = {
  comment: string | null
  useName: boolean
  type?: string
}

// Returns signature ID
export type CreateSignatureResDto = number
