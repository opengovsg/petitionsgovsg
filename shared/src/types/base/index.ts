export * from './signature'
export * from './post'
export * from './user'
export * from './addressee'
