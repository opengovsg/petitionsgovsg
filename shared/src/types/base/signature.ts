import { z } from 'zod'
import { BaseModel } from './common'

export const Signature = BaseModel.extend({
  comment: z.string().nullable(),
  hashedUserSgid: z.string(),
  postId: z.string(),
  fullname: z.string().nullable(),
})

export type Signature = z.infer<typeof Signature>
