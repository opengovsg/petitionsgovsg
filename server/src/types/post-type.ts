import { Post } from '~shared/types/base'

export type PostEditType = {
  id: number
  userid: number
  summary: string
  title: string
}

export type UpdatePostRequestDto = Pick<Post, 'title' | 'summary'>
