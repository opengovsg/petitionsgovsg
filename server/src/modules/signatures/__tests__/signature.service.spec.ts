import { okAsync } from 'neverthrow'
import { Sequelize } from 'sequelize'
import { Post, PostStatus, User, Signature } from '~shared/types/base'
import { ModelDef } from '../../../types/sequelize'
import {
  createTestDatabase,
  getModelDef,
  ModelName,
} from '../../../util/jest-db'
import { SignatureService } from '../signature.service'

describe('AnswersService', () => {
  let user: User
  let post: Post
  let signature: Signature

  let db: Sequelize
  let Post: ModelDef<Post>
  let Signature: ModelDef<Signature>
  let service: SignatureService

  const searchSyncService = {
    createPost: jest.fn(),
    updatePost: jest.fn(),
  }

  beforeAll(async () => {
    db = await createTestDatabase()
    Post = getModelDef<Post>(db, ModelName.Post)
    Signature = getModelDef<Signature>(db, ModelName.Signature)
    service = new SignatureService({
      Post,
      Signature,
      sequelize: db,
    })

    const User = getModelDef<User>(db, ModelName.User)
    user = await User.create({
      sgid: 'u=34',
      fullname: 'Tan Ah Wee',
      email: 'tanahwee@test.gov.sg',
      active: true,
    })
  })

  beforeEach(async () => {
    await Post.destroy({ truncate: true })
    await Signature.destroy({ truncate: true })
    post = await Post.create({
      title: 'Post',
      summary: 'summary',
      userId: user.id,
      reason: null,
      request: null,
      references: null,
      status: PostStatus.Closed,
      fullname: 'Mr. Tan',
    })
    signature = await Signature.create({
      comment: 'The reason why I support',
      userId: user.id,
      postId: post.id,
      fullname: 'Mr. Tan',
    })
  })

  afterAll(async () => {
    await db.close()
  })

  describe('listAnswers', () => {
    it('returns only answers associated with specified id', async () => {
      const otherPost = await Post.create({
        title: 'Other Post',
        summary: 'summary',
        userId: user.id,
        reason: null,
        request: null,
        references: null,
        status: PostStatus.Closed,
        fullname: 'Mr. Tan',
      })
      const otherSignature = await Signature.create({
        comment: 'Other Signature Body',
        userId: user.id,
        postId: otherPost.id,
        fullname: 'Mr. Tan',
      })

      const signatures = await service.listSignatures(post.id)

      expect(signatures.length).toEqual(1)
      expect(signatures.some((a) => a.id === signature.id)).toBe(true)
      expect(signatures.some((a) => a.id === otherSignature.id)).toBe(false)
    })
  })

  describe('createSignature', () => {
    it('creates a signature for the given postId', async () => {
      const comment = 'Second Signature Body'
      const attributes = {
        comment,
        postId: post.id,
        userId: user.id,
        fullname: null,
      }

      searchSyncService.updatePost.mockResolvedValue(okAsync({}))

      const signatureId = await service.createSignature(attributes)
      const signature = await Signature.findByPk(signatureId)

      const updatedPost = await Post.findByPk(post.id)

      expect(signature).toMatchObject(attributes)
      expect(updatedPost?.status).toEqual(PostStatus.Open)
    })
  })
  describe('deleteSignature', () => {
    it('deletes the specified Signature', async () => {
      await service.deleteSignature(signature.id)

      const deletedSignature = await Signature.findByPk(signature.id)

      expect(deletedSignature).toBeNull()
    })
  })
})
