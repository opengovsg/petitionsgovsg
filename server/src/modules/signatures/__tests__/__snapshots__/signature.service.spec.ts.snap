// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignatureService checkUserHasSigned returns a signature if user has signed a petition 1`] = `
Object {
  "comment": null,
  "createdAt": Any<Date>,
  "fullname": null,
  "hashedUserSgid": "hashedUserSgid",
  "id": 4,
  "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
  "type": "support",
  "updatedAt": Any<Date>,
}
`;

exports[`SignatureService checkUserHasSigned returns null if user has not signed a petition 1`] = `null`;

exports[`SignatureService createSignature returns a signature for specified post id 1`] = `
Object {
  "comment": "I want to support this cause.",
  "createdAt": Any<Date>,
  "fullname": null,
  "hashedUserSgid": "$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW",
  "id": 4,
  "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
  "type": "support",
  "updatedAt": Any<Date>,
}
`;

exports[`SignatureService listSignatures returns only signatures associated with specified post id 1`] = `
Array [
  Object {
    "comment": null,
    "createdAt": 2022-01-01T00:00:00.000Z,
    "fullname": null,
    "hashedUserSgid": "1234",
    "id": 1,
    "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
    "type": "support",
    "updatedAt": 2022-01-01T00:00:00.000Z,
  },
  Object {
    "comment": null,
    "createdAt": 2022-01-01T00:00:00.000Z,
    "fullname": "John Doe",
    "hashedUserSgid": "5678",
    "id": 2,
    "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
    "type": "support",
    "updatedAt": 2022-01-01T00:00:00.000Z,
  },
  Object {
    "comment": "I support this cause!",
    "createdAt": 2022-01-01T00:00:00.000Z,
    "fullname": null,
    "hashedUserSgid": "8765",
    "id": 3,
    "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
    "type": "support",
    "updatedAt": 2022-01-01T00:00:00.000Z,
  },
]
`;
